### ISSUE TITLE
feat: ion-infinite-scroll: 'skip' offset required

### ISSUE THREAD

**Issue by tohagan** [driftyco/ionic#3849]
**Type**: <span ionic-type>feat</span>

**Platform**: <span ionic-platform>all</span>  

<span ionic-description>I've recently migrated from several years coding Sencha Touch to Ionic and for the most part I'm preferring Ionic but for many business apps containing large numbers of database rows (we're talking 100,000 or more) this missing feature is a *show stopper*.  Currently it's the only outstanding reason why I can't yet convince my employer to switch to Ionic.

`ion-infinite-scroll` assumes that an array containing all loaded data is sufficient but IMHO this model breaks because for datasets with large numbers of rows invariably we run out of memory. In many business apps we need to allow uses to change to a large offset in the viewed dataset.  This would crash in your current implementation. A more useful model would be to assume that the scrollable data is a *sliding window* of the larger dataset rather than assume it is always anchored to the start of the dataset.

You could easily do this by adding an *offset* variable to your existing model and then treat the array as starting at this offset (defaulting the offset to 0 for backward compatibility). To support smooth scrolling this data window needs to be larger than the actual data displayed on the screen. It also needs to support server side loading when scrolling either forward (currently supported) or *backwards* (not yet supported).  This will reduce the client side memory and avoid the "out of memory" scenario. 

P.S. Love the other newest scroll features. Some of these options and optimisations are a step ahead of Sencha.</span>

<span is-issue-template></span>

**Comment by mhartington** [comment]
What you’re looking to achieve could be accomplished out of the box by using infinite scroll with collection repeat. you could use the scroll delegate to jump to certain positions if necessary as well

**Comment by tohagan** [comment]
Ok then I wonder if there's a code sample somewhere. Since I obviously missed the connection, I'm guessing others are too.

**Comment by tohagan** [comment]
I found this article and will follow the discussion thread mentioned:
  - http://www.thomasmaximini.com/2015/01/12/improving-scroll-performance-for-lists-in-ionic-framework.html

Still really needs a code sample to connect the dots. Thanks!

**Comment by mhartington** [comment]
The code would essentially be same as if you were using ng-repeat. Except you would swap out ng-repeat for collection-repeat.


```html
<ion-list>
  <ion-item collection-repeat="c in contacts">
    {{c.name}}
  </ion-item
  <ion-infinite-scroll on-infinite="loadMore()"></ion-infinite-scroll>
</ion-list>
```

**Comment by tohagan** [comment]
Thanks for this Mike.  I see that Raymond Camdon [covers this](http://www.raymondcamden.com/2014/07/10/Video-example-collectionrepeat-performance-in-Ionic)

The bit I'm now missing is how do I get the collection to maintain a window of data from server as the user scrolls down *and up* ?  I don't see what disposes of records from the memory of the collection that are no longer visible. I understand that the DOM rows are recycled by collection-repeat - but I want to only keep a slice of the data records in memory as well. This rather important when you have millions of rows and need to be able to allow the user to jump to any offset in that dataset (and do server side sorting/filtering). loadMore() suggests that it will still only *add* but not remove records from the collection (right?) so it's not a sliding window.

**Comment by villeapvirtanen** [comment]
I really miss these features too, and I think this is really essential - please be aware that the real world needs infinite scrolling as well as infinite sliding box to all directions.

For an example we currently have a view where user can see the items for the current month, left or right swipe go to previous or next month, and the screen should be infinitely scrolling, so that when you reach the bottom of the current month, the previous is loaded. And when you try to scroll even more up, the next should be loaded. So essentially this is a list scrolling in time, and left right swipe dedicated to quick month jumps.

Once user edits or adds a record and is done, we should land him/her to the correct place in the list, requiring ability to jump to correct month. (The user has the ability to select year and month when editing the record.)

Initially I thought that this is quick to implement using slidebox and initinite list, but boy I was wrong.

Shortly put: we need 
 * sliding-window-scroll
 * sliding-window-slide-box

Those should provide a window to the data, with the ability for the app dev code to signal the item that we've reached the first record / last record. You should also let the client code to decide how many records to keep in memory and how to do the caching etc.

Another way to fix this would be to offer an easy way to just detect when user has reached the start / end of the list in addition to easily call animations. (How can I declare that on left swipe I'd like to throw the view out of the left side screen, populate some data, and bring the view in from the right side?)

After this rant, thank you for the nice framework, I hope you get these things quickly resolved!

**Comment by Ionitron** [comment]
Greetings @tohagan!

My sensors indicate a reply was requested, but as we have not received one yet, I am using my robot powers to close this issue. However, if you are still experiencing this issue, please feel free to reopen it by creating a new issue, and include any examples or necessary information.

Thank you for allowing me to assist you.

**Comment by Ionitron** [comment]
Greetings @tohagan!

My sensors indicate a reply was requested, but as we have not received one yet, I am using my robot powers to close this issue. However, if you are still experiencing this issue, please feel free to reopen it by creating a new issue, and include any examples or necessary information.

Thank you for allowing me to assist you.
